### PR TITLE
feat(views): Recent Workspaces Panel

### PR DESCRIPTION
**Pull request by jonathanyeung**
## feat(views): Recent Workspaces Panel

Adds a panel displaying recently opened Dendron workspaces.  This panel will show the last 5 opened Dendron workspaces.  If a user has never opened a workspace before, then they'll get a message to get started (see below). Clicking on an item will open the workspace. 

This only shows up if not in a currently active Dendron workspace. This is to prevent clutter in the side bar.  If a user wants to manage multiple projects, then I think we should point them to the Project Manager extension, which is a fully featured version of the same concept.

I also had to make some adjustments in LaunchTutorial code + some telemetry tweaks.

---

## Dendron Extended PR Checklist

- NOTE: the links don't work. you'll need to go into the wiki and use lookup to find the note until we fix some issues in the markdown export

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
Added to 'Upcoming'
- [x] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [x] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [x] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

## Docs
- [ ] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [ ] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### CODE REVIEW

File: packages/plugin-core/src/constants.ts [hunk 1]
```diff
@@ -872,15 +897,32 @@ export const DENDRON_COMMANDS: { [key: string]: CommandEntry } = {
...
+    key: "dendron.launchTutorialWorkspace",
+    title: `${CMD_PREFIX} Launch Tutorial Workspace`,
+    when: "false",
+  },
   OPEN_BACKUP: {
     key: "dendron.openBackup",
     title: `${CMD_PREFIX} Open Backup`,
     when: DendronContext.PLUGIN_ACTIVE,
   },
+  INSTRUMENTED_WRAPPER_COMMAND: {
```
Comment (kevinslin) [inline packages/plugin-core/src/constants.ts:921]: can you add a comment to what this means?

File: packages/plugin-core/src/commands/LaunchTutorialCommand.ts [hunk 2]
```diff
@@ -10,6 +10,10 @@ type CommandInput = {};
 
 type CommandOutput = void;
 
+/**
+ * This command is a bit of a misnomer - it actually launches the welcome
+ * webview page
+ */
 export class LaunchTutorialCommand extends BasicCommand<
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/LaunchTutorialCommand.ts:17]: can we rename it in that case?

File: packages/plugin-core/src/commands/InstrumentedWrapperCommand.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+/**
+ * This command is a simple wrapper around commands, but this wrapper also fires
+ * a telemetry event when it is invoked. This is intended to be used as a
+ * wrapper around built-in VSCode commands that need to be invoked via a command
+ * URI, such as within webviews or in TreeView items, but where we still want a
+ * telemetry event to get reported.
+ */
+export class InstrumentedWrapperCommand extends BasicCommand<
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/InstrumentedWrapperCommand.ts:30]: can we do an async and add it to the instrumentation cookbook after the pr has been merged?

File: packages/plugin-core/src/features/RecentWorkspacesTreeview.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      command: {
+        title: "Open Workspace",
+        command: "vscode.openFolder",
+        arguments: [Uri.file(element.fsPath)],
+      },
+      event: VSCodeEvents.RecentWorkspacesPanelUsed,
+    };
+
+    return {
+      label: element.fsPath,
```
Comment (kevinslin) [inline packages/plugin-core/src/features/RecentWorkspacesTreeview.ts:37]: in terms of ergonomics, the whole path to the note overflows the sidebar which is not that great: https://www.loom.com/i/b8d556dc80584b0da3caf2b2b8073c1f
can we list vaults by their names instead (and tooltip to the full path)?

File: packages/engine-server/src/metadata/service.ts [hunk 5]
```diff
@@ -233,6 +237,10 @@ export class MetadataService {
     return this.getMeta().priorTools;
   }
 
+  get RecentWorkspaces(): string[] | undefined {
```
Comment (kevinslin) [inline packages/engine-server/src/metadata/service.ts:240]: we have a mix of `get {Prop}` and `get{Prop}` calls here (and everywhere). can we have an async to standardize? ()

**Review by kevinslin — CHANGES_REQUESTED** [review]
looking forward to this feature. some callouts:
- we are calling this the `recent workspaces panel` - moving forward with self contained vaults, are we still using that terminology even if everything is a vault? should this just be recent vaults?
- there is a difference btw vaults that exist independently and vaults that exist as dependencies to the current vault. we should think about the UX for this even if we're not implementing it now. currently, the only way to switch between vaults consistently is to look up the `root` note of a particular vault. does it make sense to show `dependent vaults` here?
- in terms of clutter on the sidebar, can we not just minimize the `recent` panel if in an workspace vs hiding it completely?

File: packages/plugin-core/src/features/RecentWorkspacesTreeview.ts (hunk 4 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/features/RecentWorkspacesTreeview.ts:37]: I considered this but thought it'd be more informative to a new user to list the file path as opposed to the vault name - the main intention here is to make sure new users can get back into Dendron if they had closed it out. Having this be a convenience to opening vaults/workspaces for experience users is a secondary goal.

File: packages/plugin-core/src/commands/InstrumentedWrapperCommand.ts (hunk 3 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/InstrumentedWrapperCommand.ts:30]: sure

File: packages/plugin-core/src/constants.ts (hunk 1 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/constants.ts:921]: The comment is in `InstrumentedWrapperCommand.ts`

File: packages/plugin-core/src/commands/LaunchTutorialCommand.ts (hunk 2 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/LaunchTutorialCommand.ts:17]: sure - at first I left it b/c I thought it would interfere with our existing charts, but it looks like we don't look at this.

**Comment by jonathanyeung** [comment]
> we are calling this the recent workspaces panel - moving forward with self contained vaults, are we still using that terminology even if everything is a vault? should this just be recent vaults?

> there is a difference btw vaults that exist independently and vaults that exist as dependencies to the current vault. we should think about the UX for this even if we're not implementing it now. currently, the only way to switch between vaults consistently is to look up the root note of a particular vault. does it make sense to show dependent vaults here?

I think maybe we're slightly misaligned on the spec for this particular feature. The main intention of _this_ change is to make sure new users can get back into Dendron if they had closed it out. Having this be a convenience to opening vaults/workspaces for experienced users is a secondary goal.

With this in mind, the list of things that appear in this panel are either workspace files or workspace folders that Dendron recognizes.  So for accuracy, I think this should be called 'Recent Workspaces' and not 'Recent Vaults'.  This list represents the VS Code concept of workspace/folder, not Dendron Vaults.

> in terms of clutter on the sidebar, can we not just minimize the recent panel if in an workspace vs hiding it completely?

For the reason listed above, I think it actually makes more sense to not be displayed when you're in a Dendron workspace.

**Review by kevinslin — APPROVED** [review]
